src/kvstore/kvstore_dist_server.h

```diff
@@ -228,11 +229,15 @@ class KVStoreDistServer {
         CHECK_EQ(req_data.vals.size(), num_rows * unit_len);
         TBlob recv_blob(data, dshape, cpu::kDevMask);  // NOLINT(*)
         NDArray recved = NDArray(recv_blob, 0);
-        // TODO(haibin) temporarily initialized as dense NDArray. We need inplace operator
-        // support for rowsparse ndarrays. And after that `stored` should be initialized as
-        // RowSparse NDArray
         stored = NDArray(kRowSparseStorage, dshape, Context());
-        CopyFromTo(recved, &stored, 0);
+        Engine::Get()->PushSync([recved, stored](RunContext ctx) {
+            NDArray rsp = stored;
+            stored.CheckAndAlloc({mshadow::Shape1(recved.shape()[0])});
+            mshadow::Stream<cpu> *s = ctx.get_stream<cpu>();
+            op::PopulateFullIdxRspImpl(s, &rsp);
+            mshadow::Copy(recved.data().FlatTo1D<cpu, float>(), rsp.data().FlatTo1D<cpu, float>());
```

Comment:
Stream is optional arg. 
https://github.com/dmlc/mshadow/blob/master/mshadow/tensor.h#L738
Everything on kvstore server is cpu only. I can add that to make it explicit.